client/packages/system/src/Item/context/useUnitVariant.ts

```diff
@@ -1,31 +1,40 @@
 import { useUnitVariantList } from '../api';
 import { create } from 'zustand';
-import { UnitVariantNode } from '@common/types';
+import { UnitVariantNode, VariantNode } from '@common/types';
 import { NumUtils, isEqual } from '@common/utils';
 import { useEffect } from 'react';
 
+type UserSelectedVariants = {
+  [itemId: string]: /* userSelectedVariantId */ string;
+};
 interface UnitState {
   // From back end
   items: {
     [itemId: string]: UnitVariantNode;
   };
+  userSelectedVariants: UserSelectedVariants;
+  setUserSelectedVariant: (_: { itemId: string; variantId: string }) => void;
   // Should be called on startup when fetching multi unit variants
   setItems: (newItems: UnitVariantNode[]) => void;
 }
 
 const useUnitStore = create<UnitState>(set => {
   return {
     items: {},
-    userSelected: {},
+    userSelectedVariants: {},
     // TODO add user selected
-
+    setUserSelectedVariant: ({ itemId, variantId }) =>
+      set(({ userSelectedVariants, items }) => ({
+        items,
+        userSelectedVariants: { ...userSelectedVariants, [itemId]: variantId },
+      })),
     setItems: newItems =>
-      set(() => {
+      set(({ userSelectedVariants }) => {
```

Comment:
some thoughts on the `items` assignment, which I must have missed previously:

I would usually spread first and assign second, so that later items overwrite earlier ones if the id is the same, shouldn't be an issue here, it's just that usually the object creation like this is to patch an object and overwrite entries with new values

```
(acc, item) => ({ ...acc, [item.itemId]: item }),
```

I thought 'surely we've done this before..' and 'why isn't this a helper method?'
Maybe add to `ObjUtils` or `ArrayUtils`?

I also wondered about how efficient the reduce approach is, given that it has to create a new object for each array entry. So I tried this approach:

```
        const items = {};
        newItems.forEach(item => {
          Object.defineProperty(items, item.itemId, { value: item });
        });
```

and ran both options over 5,000 very simple item objects (just an itemId and numeric value) giving this result:

reduce: 1048.97705078125 ms
defineProperty: 0.751953125 ms

```

const items = Array.from(new Array(5000), (_val, idx) => ({itemId: idx, value: idx}));
console.time('reduce');
        const i2 = items.reduce(
          (acc, item) => ({ [item.itemId]: item, ...acc }),
          {}
        );
console.timeEnd('reduce');

console.time('defineProperty');
const i1 = {};
items.forEach(item => {
  Object.defineProperty(i1, item.itemId, { value: item });
});
console.timeEnd('defineProperty');
```

I think it's worth changing! would be good to have as a utility function too in this case, as I expect that I'll try and write a version with reduce next time I need this

Reply:
Thanks Mark !

Added keyBy to utils and used in useVariant: 923e1080b57fac050340fc2557fea023906b0d2b

Benched keyBy just in case (btw thanks for the really cool bench example with console.time):
![Screenshot 2023-09-05 at 12 34 17 PM](https://github.com/openmsupply/open-msupply/assets/26194949/0ba031de-f935-4bd6-918f-071017c654dd)

Looks like lodash uses it's own optimised impleentation
https://github.com/lodash/lodash/blob/2da024c3b4f9947a48517639de7560457cd4ec6c/.internal/arrayReduce.js#L12